pkg/clusterapi/aws/actuator.go

```diff
@@ -367,35 +339,114 @@ func (a *Actuator) DeleteMachine(machine *clusterv1.Machine) error {
 		return fmt.Errorf("error getting EC2 client: %v", err)
 	}
 
-	terminateInstancesRequest := &ec2.TerminateInstancesInput{
-		InstanceIds: []*string{
-			aws.String(instanceID),
-		},
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return err
+	}
+	if len(instances) == 0 {
+		mLog.Warnf("no instances found to delete for machine")
+		return nil
 	}
-	terminateInstancesResult, err := client.TerminateInstances(terminateInstancesRequest)
+
+	instanceIDs := []*string{}
+	for _, instance := range instances {
+		mLog.WithField("instanceID", *instance.InstanceId).Infof("deleting AWS instance")
+		instanceIDs = append(instanceIDs, instance.InstanceId)
+	}
+
+	return a.terminateInstances(client, instanceIDs, mLog)
+}
+
+// Update attempts to sync machine state with an existing instance. Today this just updates status
+// for details that may have changed. (IPs and hostnames) We do not currently support making any
+// changes to actual machines in AWS. Instead these will be replaced via MachineDeployments.
+func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	mLog := clustoplog.WithMachine(a.logger, machine)
+	mLog.Debugf("updating machine")
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
+	if err != nil {
+		return err
+	}
+
+	if clusterSpec.Hardware.AWS == nil {
+		return fmt.Errorf("Cluster does not contain an AWS hardware spec")
+	}
+
+	coMachineSetSpec, err := controller.MachineSetSpecFromClusterAPIMachineSpec(&machine.Spec)
 	if err != nil {
-		// If the instance no longer exists, we need to make sure we don't infinitely
-		// block deletion of the Machine in etcd.
-		if awsErr, ok := err.(awserr.Error); ok {
-			if awsErr.Code() == ec2InstanceIDNotFoundCode {
-				mLog.WithField("instanceID", instanceID).Warnf("AWS instance no longer exists")
-				return nil
+		return err
+	}
+
+	region := clusterSpec.Hardware.AWS.Region
+	mLog.WithField("region", region).Debugf("obtaining EC2 client for region")
+	client, err := a.ec2Client(coMachineSetSpec, machine.Namespace, region)
+	if err != nil {
+		return fmt.Errorf("unable to obtain EC2 client: %v", err)
+	}
+
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return err
+	}
+
+	if len(instances) == 0 {
+		// Parent controller should prevent this from ever happening by calling Exists and then Create,
+		// but instance could be deleted between the two calls.
+		return fmt.Errorf("attempted to update machine but no instances found")
+	}
+
+	// In very unusual circumstances, there could be more than one machine running matching this
+	// machine name and cluster ID. In this scenario we will keep the newest, and delete all others.
+	var newestInstance *ec2.Instance
+	if len(instances) > 1 {
+
+		// Determine which instance is the newest:
+		for _, instance := range instances {
+			mLog.WithFields(log.Fields{
```

Comment:
Thinking that this might be simpler with a sort on LaunchTime

Reply:
Probably a good idea. I've got a followup which refactors this a bit for re-use in the AWS ELB controller, if it's ok I'll incorporate a sort there in next PR.

Reply:
👍